### PR TITLE
Allow HiveQueryRunner to persist data to a custom path

### PR DESCRIPTION
**Pull request by yingsu00**
Add an optional argument to HiveQueryRunner to provide a path to read
    and write the data and metadata to. The path must point to a writable
    directory. On first run, HiveQueryRunner will create tpch and
    tpch_bucketed tables and store them in the specified directory. On
    subsequent runs using the same path, HiveQueryRunner will re-use the
    data created before.

    If no argument is provided, HiveQueryRunner will create a temporary
    directory to store data and delete it before exiting.

### CODE REVIEW

File: presto-hive/src/test/java/com/facebook/presto/hive/HiveQueryRunner.java [hunk 1]
```diff
@@ -75,8 +75,20 @@ public static DistributedQueryRunner createQueryRunner(Iterable<TpchTable<?>> ta
         return createQueryRunner(tables, extraProperties, "sql-standard", ImmutableMap.of());
     }
 
+    public static DistributedQueryRunner createQueryRunner(Iterable<TpchTable<?>> tables, Map<String, String> extraProperties, Optional<String> baseDirString)
```
Comment (electrum) [inline presto-hive/src/test/java/com/facebook/presto/hive/HiveQueryRunner.java]: Use `File` to pass arguments. Converting from a string to a path should be a concern for the caller. APIs in general should use appropriate types rather than strings.

File: presto-hive/src/test/java/com/facebook/presto/hive/HiveQueryRunner.java [hunk 2]
```diff
@@ -75,8 +75,20 @@ public static DistributedQueryRunner createQueryRunner(Iterable<TpchTable<?>> ta
...
+        return createQueryRunner(tables, extraProperties, "sql-standard", ImmutableMap.of(), baseDirString);
+    }
+
     public static DistributedQueryRunner createQueryRunner(Iterable<TpchTable<?>> tables, Map<String, String> extraProperties, String security, Map<String, String> extraHiveProperties)
             throws Exception
+    {
+        return createQueryRunner(tables, extraProperties, security, extraHiveProperties, Optional.empty());
+    }
+
+    public static DistributedQueryRunner createQueryRunner(Iterable<TpchTable<?>> tables, Map<String, String> extraProperties, String security, Map<String, String> extraHiveProperties, Optional<String> baseDirString)
```
Comment (electrum) [inline presto-hive/src/test/java/com/facebook/presto/hive/HiveQueryRunner.java]: Use `File` to pass arguments. Converting from a string to a path should be a concern for the caller. APIs in general should use appropriate types rather than strings.

File: presto-hive/src/test/java/com/facebook/presto/hive/HiveQueryRunner.java (hunk 1 above)
Comment (electrum) [inline presto-hive/src/test/java/com/facebook/presto/hive/HiveQueryRunner.java]: Rather than adding a new method, change the one above, as it's only called by the `main` method (and would be unused after this change).

File: presto-hive/src/test/java/com/facebook/presto/hive/HiveQueryRunner.java [hunk 3]
```diff
@@ -75,8 +75,20 @@ public static DistributedQueryRunner createQueryRunner(Iterable<TpchTable<?>> ta
         return createQueryRunner(tables, extraProperties, "sql-standard", ImmutableMap.of());
     }
 
+    public static DistributedQueryRunner createQueryRunner(Iterable<TpchTable<?>> tables, Map<String, String> extraProperties, Optional<String> baseDirString)
+            throws Exception
+    {
+        return createQueryRunner(tables, extraProperties, "sql-standard", ImmutableMap.of(), baseDirString);
+    }
+
     public static DistributedQueryRunner createQueryRunner(Iterable<TpchTable<?>> tables, Map<String, String> extraProperties, String security, Map<String, String> extraHiveProperties)
```
Comment (electrum) [inline presto-hive/src/test/java/com/facebook/presto/hive/HiveQueryRunner.java]: Let's also not preserve this method, as it's only used in one place.

**Comment by electrum** [comment]
Another approach is to allow configuring `baseDir` of the coordinator. That would allow preserving data for any query runner.

**Comment by yingsu00** [comment]
> Another approach is to allow configuring `baseDir` of the coordinator. That would allow preserving data for any query runner.

@electrum Are we talking about using a session property to set up the baseDir? In that case the user would have to set the session property when the server is started, then restart the server to honor this property. If we pass in the baseDir from program arguments we don't need to restart the server and it's easier to change it. However if the user forgets the baseDir (e.g. lost the run configuration) then the session property way is better. I'm ok with either way. Which way do you prefer?

**Comment by yingsu00** [comment]
@electrum Hi David I updated the commit with your comments addressed. I'm still passing in a program argument from main(). Please let me know if you have preference using a session property instead. Thanks!

**Comment by electrum** [comment]
The base directory would be configured as a constructor argument to `TestingPrestoServer`. The difference is in configuring the directory for the entire server vs just the Hive connector.

File: presto-hive/src/test/java/com/facebook/presto/hive/HiveQueryRunner.java [hunk 4]
```diff
@@ -200,7 +210,16 @@ public static void main(String[] args)
     {
         // You need to add "--user user" to your CLI for your queries to work
         Logging.initialize();
-        DistributedQueryRunner queryRunner = createQueryRunner(TpchTable.getTables(), ImmutableMap.of("http-server.http.port", "8080"));
+
+        File baseDirFile = null;
+        if (args.length > 0) {
+            if (args.length != 1) {
+                log.error("Only one String argument (baseDir) is allowed. It must be a URI the user has write permission.");
```
Comment (electrum) [inline presto-hive/src/test/java/com/facebook/presto/hive/HiveQueryRunner.java]:
```java
System.err("usage: HiveQueryRunner [baseDir]\n");
System.exit(1);

File: presto-hive/src/test/java/com/facebook/presto/hive/HiveQueryRunner.java [hunk 5]
```diff
@@ -200,7 +210,16 @@ public static void main(String[] args)
     {
         // You need to add "--user user" to your CLI for your queries to work
         Logging.initialize();
-        DistributedQueryRunner queryRunner = createQueryRunner(TpchTable.getTables(), ImmutableMap.of("http-server.http.port", "8080"));
+
+        File baseDirFile = null;
```
Comment (electrum) [inline presto-hive/src/test/java/com/facebook/presto/hive/HiveQueryRunner.java]: Declare this as `Optional` rather than using null.

File: presto-hive/src/test/java/com/facebook/presto/hive/HiveQueryRunner.java [hunk 6]
```diff
@@ -200,7 +210,16 @@ public static void main(String[] args)
...
+        if (args.length > 0) {
+            if (args.length != 1) {
+                log.error("Only one String argument (baseDir) is allowed. It must be a URI the user has write permission.");
+                return;
+            }
+            baseDirFile = new File(args[0]);
+        }
+        DistributedQueryRunner queryRunner = createQueryRunner(TpchTable.getTables(), ImmutableMap.of("http-server.http.port", "8080"), Optional.ofNullable(baseDirFile));
```
Comment (electrum) [inline presto-hive/src/test/java/com/facebook/presto/hive/HiveQueryRunner.java]: Add a newline to separate argument processing from query runner creation.

**Comment by yingsu00** [comment]
@electrum Hi David I updated the PR using the way you suggested. Two questions 1) I added one new constructors in DistributedQueryRunner. However I saw the constructors were marked as deprecated. I didn't know why but also added the @Deprecated annotation for the new constructor 2) I didn't add a new createTestingPrestoServer function but just modified it to take the baseDirPath. So the workers are updated with this baseDirPath too. I think this is fine because the baseDataDir is not used for workers. Please let me know you think otherwise. Thanks!

File: presto-main/src/main/java/com/facebook/presto/server/testing/TestingPrestoServer.java [hunk 7]
```diff
@@ -171,16 +171,34 @@ public TestingPrestoServer(List<Module> additionalModules)
         this(true, ImmutableMap.of(), null, null, new SqlParserOptions(), additionalModules);
     }
 
+    public TestingPrestoServer(boolean coordinator,
+                               Map<String, String> properties,
+                               String environment,
+                               URI discoveryUri,
+                               SqlParserOptions parserOptions,
+                               List<Module> additionalModules)
+            throws Exception
```
Comment (dain) [inline presto-main/src/main/java/com/facebook/presto/server/testing/TestingPrestoServer.java]: Formatting is off

File: presto-main/src/main/java/com/facebook/presto/server/testing/TestingPrestoServer.java [hunk 8]
```diff
@@ -171,16 +171,34 @@ public TestingPrestoServer(List<Module> additionalModules)
...
             SqlParserOptions parserOptions,
-            List<Module> additionalModules)
+            List<Module> additionalModules,
+            Optional<Path> baseDataDirPath)
             throws Exception
     {
         this.coordinator = coordinator;
-        baseDataDir = Files.createTempDirectory("PrestoTest");
+
+        if (!baseDataDirPath.isPresent()) {
```
Comment (dain) [inline presto-main/src/main/java/com/facebook/presto/server/testing/TestingPrestoServer.java]: Use the `Optional` `map` and `orElse` apis instead of an if/else block

**Comment by yingsu00** [comment]
Hi @dain updated the PR with your comments fixed. Thanks for reviewing! Happy new year.

File: presto-main/src/main/java/com/facebook/presto/server/testing/TestingPrestoServer.java [hunk 9]
```diff
@@ -178,9 +178,22 @@ public TestingPrestoServer(boolean coordinator,
...
+            URI discoveryUri,
+            SqlParserOptions parserOptions,
+            List<Module> additionalModules,
+            Optional<Path> baseDataDirPath)
+            throws Exception
     {
         this.coordinator = coordinator;
-        baseDataDir = Files.createTempDirectory("PrestoTest");
+
+        baseDataDir = baseDataDirPath.orElse(Files.createTempDirectory("PrestoTest"));
```
Comment (dain) [inline presto-main/src/main/java/com/facebook/presto/server/testing/TestingPrestoServer.java]: you should use the lambda version of `orElse` to avoid the creation of the temp directory.  Also I'd static import that method.

I'd also perform some basic validation on the path since it is now user supplied.  If it already exists, it should be a directory.  If not, either throw an exception or possibly create the directory (not sure which is more desirable).

Comment (electrum) [reply presto-main/src/main/java/com/facebook/presto/server/testing/TestingPrestoServer.java]: The cleanup code also needs to be changed not to delete the directory if it is user provided.

File: presto-main/src/main/java/com/facebook/presto/server/testing/TestingPrestoServer.java (hunk 8 above)
Comment (yingsu00) [reply presto-main/src/main/java/com/facebook/presto/server/testing/TestingPrestoServer.java]: @dain createTempDirectory("PrestoTest") throws IOException, and Optional.orElseGet() takes a Supplier<? extends T> parameter which doesn't support throwing checked exceptions. To get around it we need to provide a lambda wrapper of Supplier but then it loses the succinctness of using lambda here. Since we also need to remember isBaseDataDirUserSupplied, I reverted back to the original way as I think it's easier to read and understand. Please let me know if there are other better ways.

File: presto-main/src/main/java/com/facebook/presto/server/testing/TestingPrestoServer.java (hunk 9 above)
Comment (yingsu00) [reply presto-main/src/main/java/com/facebook/presto/server/testing/TestingPrestoServer.java]: @dain The validation was done in RawLocalFileSystem.mkdirsWithOptionalPermission() and ChecksumFileSystem.create() when creating database. In that case exceptions would be thrown on invalid user supplied path when creating database. I didn't include the validation in HiveQueryRunner.main() originally because I thought the logic would be repeated twice. But doing it in HiveQueryRunner would allow invalid path to fail sooner than later in createDatabase(), so I copied the validation logic to HiveQueryRunner.main() in the new commit.

> I'd also perform some basic validation on the path since it is now user supplied. If it already exists, it should be a directory. If not, either throw an exception or possibly create the directory (not sure which is more desirable).

**Comment by yingsu00** [comment]
@electrum @dain Any thought on the orElseGet()? I have addressed your other comments. Could you please take a look again? Thanks!

File: presto-hive/src/test/java/com/facebook/presto/hive/HiveQueryRunner.java [hunk 10]
```diff
@@ -200,7 +208,41 @@ public static void main(String[] args)
     {
         // You need to add "--user user" to your CLI for your queries to work
         Logging.initialize();
-        DistributedQueryRunner queryRunner = createQueryRunner(TpchTable.getTables(), ImmutableMap.of("http-server.http.port", "8080"));
+
+        Optional<Path> baseDirPath = Optional.empty();
```
Comment (mbasmanova) [inline presto-hive/src/test/java/com/facebook/presto/hive/HiveQueryRunner.java]: s/baseDirPath/baseDataDir throughout

File: presto-hive/src/test/java/com/facebook/presto/hive/HiveQueryRunner.java [hunk 11]
```diff
@@ -66,21 +68,21 @@ public static DistributedQueryRunner createQueryRunner(TpchTable<?>... tables)
...
-        return createQueryRunner(tables, ImmutableMap.of());
+        return createQueryRunner(tables, ImmutableMap.of(), Optional.empty());
     }
 
-    public static DistributedQueryRunner createQueryRunner(Iterable<TpchTable<?>> tables, Map<String, String> extraProperties)
+    public static DistributedQueryRunner createQueryRunner(Iterable<TpchTable<?>> tables, Map<String, String> extraProperties, Optional<Path> baseDirPath)
             throws Exception
     {
-        return createQueryRunner(tables, extraProperties, "sql-standard", ImmutableMap.of());
+        return createQueryRunner(tables, extraProperties, "sql-standard", ImmutableMap.of(), baseDirPath);
```
Comment (nezihyigitbasi) [inline presto-hive/src/test/java/com/facebook/presto/hive/HiveQueryRunner.java]: - In other parts of the code we use `baseDataDir`, so we should be consistent.
- In general we don't use abbreviations in Presto, so I think we can do the same here and rename these variables to `baseDataDirectory`.

File: presto-hive/src/test/java/com/facebook/presto/hive/HiveQueryRunner.java [hunk 12]
```diff
@@ -93,8 +95,7 @@ public static DistributedQueryRunner createQueryRunner(Iterable<TpchTable<?>> ta
             HdfsEnvironment hdfsEnvironment = new HdfsEnvironment(hdfsConfiguration, hiveClientConfig, new NoHdfsAuthentication());
 
             FileHiveMetastore metastore = new FileHiveMetastore(hdfsEnvironment, baseDir.toURI().toString(), "test");
-            metastore.createDatabase(createDatabaseMetastoreObject(TPCH_SCHEMA));
-            metastore.createDatabase(createDatabaseMetastoreObject(TPCH_BUCKETED_SCHEMA));
+
```
Comment (nezihyigitbasi) [inline presto-hive/src/test/java/com/facebook/presto/hive/HiveQueryRunner.java]: Looks like an unnecessary whitespace change.

File: presto-hive/src/test/java/com/facebook/presto/hive/HiveQueryRunner.java [hunk 13]
```diff
@@ -200,7 +208,41 @@ public static void main(String[] args)
     {
         // You need to add "--user user" to your CLI for your queries to work
         Logging.initialize();
-        DistributedQueryRunner queryRunner = createQueryRunner(TpchTable.getTables(), ImmutableMap.of("http-server.http.port", "8080"));
+
+        Optional<Path> baseDirPath = Optional.empty();
+        if (args.length > 0) {
+            if (args.length != 1) {
+                System.err.println("usage: HiveQueryRunner [baseDir]\n");
```
Comment (nezihyigitbasi) [inline presto-hive/src/test/java/com/facebook/presto/hive/HiveQueryRunner.java]: Please rename `baseDir`, it's not easy to understand what it means. Also, we use that in the error messages below, we should update those to be meaningful too.

File: presto-hive/src/test/java/com/facebook/presto/hive/HiveQueryRunner.java [hunk 14]
```diff
@@ -200,7 +208,41 @@ public static void main(String[] args)
...
+            }
+
+            File baseDirFile = new File(args[0]);
+            if (baseDirFile.exists()) {
+                if (!baseDirFile.isDirectory()) {
+                    System.err.println("Error: baseDir is not a directory: " + baseDirFile.getAbsolutePath());
+                    System.exit(1);
+                }
+                else if (!baseDirFile.canRead() || !baseDirFile.canWrite()) {
+                    System.err.println("Error: baseDir cannot be read or written: " + baseDirFile.getAbsolutePath());
```
Comment (nezihyigitbasi) [inline presto-hive/src/test/java/com/facebook/presto/hive/HiveQueryRunner.java]: `is not readable/writable`

File: presto-hive/src/test/java/com/facebook/presto/hive/HiveQueryRunner.java [hunk 15]
```diff
@@ -200,7 +208,41 @@ public static void main(String[] args)
...
+                    System.exit(1);
+                }
+            }
+            else {
+                // For user supplied path like [Path_exsits_but_cannot_read_or_write]/[paths_do_not_exist], the hadoop file system won't
```
Comment (nezihyigitbasi) [inline presto-hive/src/test/java/com/facebook/presto/hive/HiveQueryRunner.java]: nit: typo in `exsits`

File: presto-hive/src/test/java/com/facebook/presto/hive/HiveQueryRunner.java [hunk 16]
```diff
@@ -200,7 +208,41 @@ public static void main(String[] args)
...
+                // be able to create directory for it. e.g. "/aaa/bbb" is not creatable because path "/" is not writable.
+                while (!baseDirFile.exists()) {
+                    baseDirFile = baseDirFile.getParentFile();
+                }
+                if (!baseDirFile.canRead() || !baseDirFile.canWrite()) {
+                    System.err.println("Error: The parent directory of baseDir cannot be read or written: " + baseDirFile.getAbsolutePath());
```
Comment (nezihyigitbasi) [inline presto-hive/src/test/java/com/facebook/presto/hive/HiveQueryRunner.java]: `is not readable/writable`

File: presto-hive/src/test/java/com/facebook/presto/hive/HiveQueryRunner.java (hunk 11 above)
Comment (nezihyigitbasi) [reply presto-hive/src/test/java/com/facebook/presto/hive/HiveQueryRunner.java]: Also, what does `base` mean here? Why don't we rename this variable to `dataDirectory` or `localDataDirectory` if the data is only supposed to touch the local filesystem?

File: presto-main/src/main/java/com/facebook/presto/server/testing/TestingPrestoServer.java [hunk 17]
```diff
@@ -178,9 +179,29 @@ public TestingPrestoServer(boolean coordinator,
             SqlParserOptions parserOptions,
             List<Module> additionalModules)
             throws Exception
+    {
+        this(coordinator, properties, environment, discoveryUri, parserOptions, additionalModules, Optional.empty());
+    }
+
+    public TestingPrestoServer(boolean coordinator,
```
Comment (nezihyigitbasi) [inline presto-main/src/main/java/com/facebook/presto/server/testing/TestingPrestoServer.java]: nit: add a newline before `coordinator` parameter

File: presto-main/src/main/java/com/facebook/presto/server/testing/TestingPrestoServer.java [hunk 18]
```diff
@@ -178,9 +179,29 @@ public TestingPrestoServer(boolean coordinator,
...
+            SqlParserOptions parserOptions,
+            List<Module> additionalModules,
+            Optional<Path> baseDataDirPath)
+            throws Exception
     {
         this.coordinator = coordinator;
-        baseDataDir = Files.createTempDirectory("PrestoTest");
+
+        if (!baseDataDirPath.isPresent()) {
+            baseDataDir = createTempDirectory("PrestoTest");
```
Comment (nezihyigitbasi) [inline presto-main/src/main/java/com/facebook/presto/server/testing/TestingPrestoServer.java]: We can extract a method that throws an unchecked exception, and use a method reference:

```
baseDataDir = baseDataDirPath.orElseGet(TestingPrestoServer::tempDirectory);
isBaseDataDirUserSupplied = baseDataDirPath.isPresent();
```

The method will look like:

```
private static Path tempDirectory() {
    try {
        return createTempDirectory("PrestoTest");
    }
    catch (IOException e) {
        throw new UncheckedIOException(e);
    }
}
```

File: presto-tests/src/main/java/com/facebook/presto/tests/DistributedQueryRunner.java [hunk 19]
```diff
@@ -100,7 +102,14 @@ public DistributedQueryRunner(Session defaultSession, int nodeCount)
     public DistributedQueryRunner(Session defaultSession, int nodeCount, Map<String, String> extraProperties)
             throws Exception
     {
-        this(defaultSession, nodeCount, extraProperties, ImmutableMap.of(), DEFAULT_SQL_PARSER_OPTIONS, ENVIRONMENT);
+        this(defaultSession, nodeCount, extraProperties, ImmutableMap.of(), DEFAULT_SQL_PARSER_OPTIONS, ENVIRONMENT, Optional.empty());
+    }
+
+    @Deprecated
```
[the remark this answers is not in the document]
Comment (nezihyigitbasi) [reply presto-tests/src/main/java/com/facebook/presto/tests/DistributedQueryRunner.java]: We have many constructors in this class and some of them are already deprecated, let's not introduce a new deprecated one.

File: presto-hive/src/test/java/com/facebook/presto/hive/HiveQueryRunner.java (hunk 13 above)
Comment (tdcmeehan) [inline presto-hive/src/test/java/com/facebook/presto/hive/HiveQueryRunner.java]:
```suggestion
                log.error("usage: HiveQueryRunner [baseDir]\n");
```

File: presto-hive/src/test/java/com/facebook/presto/hive/HiveQueryRunner.java (hunk 11 above)
Comment (yingsu00) [reply presto-hive/src/test/java/com/facebook/presto/hive/HiveQueryRunner.java]: @nezihyigitbasi It's called baseXXX because there will be a hive_data folder created under it. In TestingPrestoServer, it used to be named as baseDataDir and getBaseDataDir():
    ```
    public Path getBaseDataDir()
    {
        return baseDataDir;
    }
```
So I suggest we still call it baseDataDir.  If we want to rename it, we can do it in a separate commit.

File: presto-main/src/main/java/com/facebook/presto/server/testing/TestingPrestoServer.java (hunk 18 above)
Comment (yingsu00) [reply presto-main/src/main/java/com/facebook/presto/server/testing/TestingPrestoServer.java]: Thanks @nezihyigitbasi . This looks more complex than writing if-else. Is this considered a better practice than writing if-else?

File: presto-main/src/main/java/com/facebook/presto/server/testing/TestingPrestoServer.java (hunk 17 above)
Comment (yingsu00) [reply presto-main/src/main/java/com/facebook/presto/server/testing/TestingPrestoServer.java]: This was copied from the other constructor. I added a new line to that one too.

File: presto-hive/src/test/java/com/facebook/presto/hive/HiveQueryRunner.java (hunk 13 above)
Comment (yingsu00) [reply presto-hive/src/test/java/com/facebook/presto/hive/HiveQueryRunner.java]: Changed to the following:
```
log.error("usage: HiveQueryRunner [baseDataDir]\n");
                log.error("       [baseDataDir] is a local directory that you want the hive_data directory will be created under.]\n");
```

**Comment by yingsu00** [comment]
> @yingsu00 Ying, thanks for extending HiveQueryRunning to re-use the data between runs and keep the data after a run for inspection. Overall the changes look good.
> 
> The commit title, "Allow HiveQueryRunner to persist data on disk", is somewhat misleading because HiveQueryRunner already persists data to disk using a temporary directory. Also, commit description doesn't clarify that after this change HiveQueryRunner will re-use the data if it already exist. Based on my understanding of this change, I'd describe it like so.
> 
> ```
> Allow HiveQueryRunner to write data to a custom path and keep that data on exit
> 
> Add an optional argument to HiveQueryRunner to provide a path to read 
> and write the data and metadata to. The path must point to a writable directory. 
> On first run, HiveQueryRunner will create tpch and tpch_bucketed tables and 
> store them in the specified directory. On subsequent runs using the same path, 
> HiveQueryRunner will re-use the data created before.
> 
> If no argument is provided, HiveQueryRunner will create a temporary
> directory to store data and delete it before exiting. 
> ```

@mbasmanova Thanks for rewriting the commit message! It looks very nice. The commit title, however, is over the 50 char limit, so I changed it to "Allow HiveQueryRunner to persist data to a custom path"

**Comment by yingsu00** [comment]
@mbasmanova @nezihyigitbasi @tdcmeehan Thanks very much for reviewing this PR! I have updated all requests, except that I kept using "baseDataDir" as the variable name. This is because it's used everywhere else even before this change, also in other query runners. We can create another commit if we really want this variable to be renamed. Please let me know if you have more comments.

**Comment by mbasmanova** [comment]
@yingsu00 Ying, Travis build is failing. Would you take a look?

File: presto-hive/src/test/java/com/facebook/presto/hive/HiveQueryRunner.java [hunk 20]
```diff
@@ -200,7 +212,43 @@ public static void main(String[] args)
...
+                while (!ancestorDirectoryFile.exists()) {
+                    ancestorDirectoryFile = ancestorDirectoryFile.getParentFile();
+                }
+                if (!ancestorDirectoryFile.canRead() || !ancestorDirectoryFile.canWrite()) {
+                    log.error("Error: The ancestor directory of " + baseDataDirFile.getAbsolutePath() + " is " + ancestorDirectoryFile.getAbsolutePath() + " and it's not readable/writable.");
+                    System.exit(1);
+                }
+            }
+
+            baseDataDir = Optional.of(Paths.get(args[0]));
```
Comment (mbasmanova) [inline presto-hive/src/test/java/com/facebook/presto/hive/HiveQueryRunner.java]: baseDataDirFile.toPath()

File: presto-tests/src/main/java/com/facebook/presto/tests/DistributedQueryRunner.java [hunk 21]
```diff
@@ -531,10 +535,16 @@ public Builder setEnvironment(String environment)
             return this;
         }
 
+        public Builder setBaseDataDir(Optional<Path> baseDataDir)
+        {
+            this.baseDataDir = baseDataDir;
```
Comment (mbasmanova) [inline presto-tests/src/main/java/com/facebook/presto/tests/DistributedQueryRunner.java]: `requireNonNull(baseDataDir, "baseDataDir is null");`

**Comment by yingsu00** [comment]
> @yingsu00 Ying, Travis build is failing. Would you take a look?

@mbasmanova Sorry I wasn't able to build locally. I just pushed again and will check again in about 10 hours. Thank you very much for reviewing!

File: presto-tests/src/main/java/com/facebook/presto/tests/DistributedQueryRunner.java [hunk 22]
```diff
@@ -531,10 +535,17 @@ public Builder setEnvironment(String environment)
             return this;
         }
 
+        public Builder setBaseDataDir(Optional<Path> baseDataDir)
+        {
+            requireNonNull(baseDataDir, "baseDataDir is null");
+            this.baseDataDir = baseDataDir;
```
Comment (mbasmanova) [inline presto-tests/src/main/java/com/facebook/presto/tests/DistributedQueryRunner.java]: @yingsu00 This can be shorted to `this.baseDataDir = requireNonNull(baseDataDir, "baseDataDir is null");`

File: presto-main/src/main/java/com/facebook/presto/server/testing/TestingPrestoServer.java (hunk 18 above)
Comment (nezihyigitbasi) [reply presto-main/src/main/java/com/facebook/presto/server/testing/TestingPrestoServer.java]: It's usually idiomatic to avoid if/else with optionals, and use functional primitives instead. This also fits better with our source code.

File: presto-tests/src/main/java/com/facebook/presto/tests/DistributedQueryRunner.java (hunk 22 above)
Comment (yingsu00) [reply presto-tests/src/main/java/com/facebook/presto/tests/DistributedQueryRunner.java]: Done. @mbasmanova
